### DISCUSSION TITLE
PyLife and GMSH workflow

### DISCUSSION THREAD

**Discussion by MiguelDLM** [boschresearch/pylife#119] [General]
Hi,

I recently started exploring fatigue and damage analysis for my research and found PyLife to be incredibly helpful. Typically, our work involves performing FEA, and we are now interested in incorporating this approach.

Here's the situation: we usually use basic FEA to compare stress distribution when applying forces to bones. We analyze the stress distribution and make inferences about the biomechanics of bones. We have been using open-source software that works with GMSH meshes. As output, we obtain .msh files with force and stress distributions (Cauchy, Von Mises, etc.). I am able to load my results with the following code:

```python
folder_path = os.path.dirname(os.path.abspath(__file__))
mesh_file = os.path.join(folder_path, 'example', 'mesh.msh')
stress_file = os.path.join(folder_path, 'example', 'stress_tensor.msh')

gmsh.initialize()
gmsh.merge(mesh_file)
gmsh.merge(stress_file)

nodeTags, nodeCoords, _ = gmsh.model.mesh.getNodes()
nodeCoords = np.array(nodeCoords).reshape(-1, 3)
nodeData = pd.DataFrame({'node_id': nodeTags, 'x': nodeCoords[:, 0], 'y': nodeCoords[:, 1], 'z': nodeCoords[:, 2]})

dataType, tags, data, time, numComp = gmsh.view.getModelData(1, 0)
svms = []
cauchy_components = []
for sig in data:
    [xx, xy, xz, yx, yy, yz, zx, zy, zz] = sig
    svm = np.sqrt(((xx - yy) ** 2 + (yy - zz) ** 2 + (zz - xx) ** 2) / 2 + 3 * (xy * xy + yz * yz + zx * zx))
    svms.append(svm)
    cauchy_components.append([xx, yy, zz, xy, xz, yz])
svms = np.array(svms)
svmData = pd.DataFrame({'mises': svms}, index=nodeTags)

cauchy_columns = ['S11', 'S22', 'S33', 'S12', 'S13', 'S23']
cauchyData = pd.DataFrame(cauchy_components, columns=cauchy_columns, index=nodeTags)

nodeData.reset_index(drop=True, inplace=True)
svmData.reset_index(drop=True, inplace=True)
cauchyData.reset_index(drop=True, inplace=True)
pyLife_mesh = pd.concat([nodeData, svmData, cauchyData], axis=1)

pyLife_mesh['mises'] /= pyLife_mesh['mises'].max()
```

However, I am unsure of the next steps. We aim to analyze failure probability, lifetime, etc., to compare the biomechanics of different species. All scenarios will be hypothetical, such as simulating walking cycles and considering different animal weights. This might be a basic question for someone experienced in this field, but I would appreciate guidance on the suggested workflow we should follow. Which concepts are relevant for this kind of research? Additionally, if you have any recommendations for relevant literature to implement this approach using PyLife, I would be grateful.

Congratulations on the development of PyLife. Tools like this are invaluable for biomechanics research and could significantly aid further studies.

**Comment by johannes-mueller** [comment]
Hi @MiguelDLM,

thanks for reaching out to us.

Basically you need three ingredients to preform a damage calculation:

* a stress distribution in a mesh – you have that
* a [load collective](https://pylife.readthedocs.io/en/stable/demos/load_collective.html) – often obtained by a rainflow analysis of a measured time signal
* a model of the strength of your material – oftentimes a [Wöhler curve](https://pylife.readthedocs.io/en/stable/demos/woehler_curve.html)

Then you can perform a damage calculation which checks if the [stress exceeds the strength](https://pylife.readthedocs.io/en/stable/demos/stress_strength.html).

As for literature, the ultimate book is [Betriebsfestigkeit (E. Haibach)](https://link.springer.com/book/10.1007/3-540-29364-7).

Feel free to provide feedback or improvements about the documentation.

cc: @DKreuter

**Reply by MiguelDLM** [reply]
Thanks for the recommendations. I will check it and once I found the answares to my doubts, I will try to write something also to help other researches on my field.

**Reply by MiguelDLM** [reply]
I noticed that for estimating the failure probability or damage calculation from an FE analysis, the stress is calculated as follows:

`mises = pyLife_mesh.groupby('element_id')[['S11', 'S22', 'S33', 'S12', 'S13', 'S23']].mean().equistress.mises()`

Does this mean that the stress used is calculated from the mean displacement of all the nodes that form the element?

In my case, my software provides the stress at the nodes, and I can extract the nodes that form each element. However, when exploring the example you provided for the "plate with hole" mesh, I noticed that the nodes shared by different elements showed different displacements, even when they are the same nodes with the same coordinates. I read that this might be because the displacement at the nodes provided by some software is the extrapolated displacement from the integration points.

Am I understanding this correctly? In this case, do I need to obtain each element's displacement at the nodes? Or is it possible to use the displacements of the nodes that form an element to estimate the mean and use this mean to estimate the stress?

**Reply by johannes-mueller** [reply]
`Sxx` is the stress, not the displacement. It is indeed different for the same node for the different elements it is involved in. The values are extrapolated from the integration points of the element.

In the code line you quoted we are averaging the stress tensor for one element by its nodes and then calculate the Mises stress of it.

**Reply by MiguelDLM** [reply]
Apologies, I sometimes mix up the terms "stress" and "displacement." I’ve already imported my data using the structure required by PyLife, and I’m now proceeding with the fatigue testing.

The ultimate goal of my work is to integrate PyLife into our workflow and current software for some upcoming publications. This will also form part of my PhD thesis. However, the process has proven to be more complex than I anticipated, especially since our research group consists of paleontologists rather than software engineers or physicists. As a result, some concepts aren’t entirely clear to me.

I am currently watching the videos from the previous workshop, and I was wondering if there will be another workshop soon. I also noticed that our location in Tübingen is quite close to your research center, so I wanted to ask if there might be any opportunity to visit for a course, receive training, or explore possible collaboration with you.

Thank you very much for your assistance.

**Reply by johannes-mueller** [reply]
Maybe it would be an option for you to look for an academic research partner doing mechanical engineering to collaborate. 

As soon as another workshop is planned, we will announce it here in the "Announcements" section.
